### ISSUE TITLE
rm4 pro

### ISSUE THREAD

**Issue by bubez81** [lbschenkel/broadlink-bridge#5]
Hello very interesting project! 
One question is the rm4 pro supported?
I received this error:
TypeError: str, bytes or bytearray expected, not rm4

**Comment by lbschenkel** [comment]
This looks like a bug. Can you please paste the whole stack trace containing the error?

**Comment by bubez81** [comment]
sure!
```
2020-11-12 10:26:55,405 INFO     Starting broadlink-bridge/0.1.5...
2020-11-12 10:26:55,406 INFO     Reading config file: config.ini
2020-11-12 10:26:55,529 INFO     Device: DC-EC-F0-A7-DF-24@192.168.1.66 has alias myalias
2020-11-12 10:26:55,529 INFO     Discovery: searching for devices for 5 seconds...
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/bin/broadlink-bridge", line 33, in <module>
    sys.exit(load_entry_point('broadlink-bridge==0.1.5', 'console_scripts', 'broadlink-bridge')())
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/cli.py", line 64, in main
    REGISTRY.discover(config.getint('discovery', 'timeout'))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 30, in discover
    dev = self._add_device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 76, in _add_device
    dev = Device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 99, in __init__
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 107, in connect
    self._dev = broadlink.hello(host=self.host)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 129, in hello
    return next(xdiscover(timeout, local_ip_address, host, port))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 160, in xdiscover
    for resp in responses:
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/device.py", line 75, in scan
    conn.sendto(packet, (discover_ip_address, discover_ip_port))
TypeError: str, bytes or bytearray expected, not rm4
```

**Comment by lbschenkel** [comment]
Unfortunately I'm not able to reproduce this. But I have a RM3, not a RM4. Can you share your config file, just in case?

**Comment by lbschenkel** [comment]
Btw: are you trying this bridge directly or via the Home Assistant add-on?

**Comment by bubez81** [comment]
directly on a Mac mini with macOS this is my config (I have a problem with the codes in Broadlink (base64) too but one issue at time:

```
[commands]
# Commands are codes that can be referred to by name.
# If an incoming code matches a command name, the command code is
# sent to the device instead.
#
# A payload can be in Broadlink (base64) format:
#philips/power-off = JgAaABweOR4bHhwdHB4dHRw6HhsdHR0dOTocAA0FAAAAAAAAAAAAAAAAAAA=
# Or it can be in Pronto hex format:
#denon/power-on = 0000 006D 0000 0020 000A 001E 000A 0046 000A 001E 000A 001E 000A 001E 000A 0046 000A 001E 000A 001E 000A 0046 000A 0046 000A 0046 000A 0046 000A 001E 000A 001E 000A 001E 000A 0677 000A 001E 000A 0046 $
# Repeats are also supported by adding a N* prefix to any code.
# For example, the code below will be sent 5 times (4 repeats):
#panasonic/power-on = 5 * 0000 0070 0000 0032 0080 0040 0010 0010 0010 0030 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0010 0030 0010 0010 0010 0010 00$
#tendadestra/up = scCwBACfBgAEBAUFBAgEBAQEBDEEBQQHBAQEBQQFBEEEBAUEBAQE/wRbBAQHBAQEBA0EnQoEBAQECAQEBGQEBAcEBwYEBAQEBPkEBQSDCQQEBQQEBAQEBAUIBAQEBAQAAY6eMhcNFw0XDQsZCxkMGQsZCxkLGQwYDBkLGQsZDBgMGRcNCxkLGRcNFw0XDRcNFw0LGQsa$
#tendadestra/down = scCwBPaeBgAEBQQAAZAELJ8yGAwXDRgMCxkMGQsZCxkLGQwZCxkLGQsZCxkLGgsZFw0LGQsZFw0XDRcNFw0XDQsZCxkLGgoaFw0LGQsZCxkXDQsaChoLGRcNCxkLGQsZF/GdMxYOFg4WDgoaChoLGQsZCxoKGgoaCxkLGQsaChoLGRcNCxkLGRcNFw4WDRcNFw0LGg$
#tendadestra/stop = scCwBACfBgAXDRcNFg0LGgsZCxkLGQsZCxoLGQsZCxkMGQsZCxkXDQsZCxoWDRcNFw4WDRcNCxoKGgoaCxkXDQsZCxoKGhYOCxkXDQsZFw0LGRcNCxkX8Z0zFg4WDhYOChoKGgsZCxkLGgoaCxkLGQsaChoLGQsZFw0QFAsaFg4WDhYOFg4WDgoaChoLGQsaFg0LGg$
#tendasinistra/up = scD8AOyeBgCdMhcNFw0XDQsZCxoLGQsZCxkLGQsaCxkLGQsZCxkLGRgMDBkLGRcNFw0XDRYOFw0KGgsZCxkLGhYNCxoLGRcNCxkLGQsaFg0XDQsaChoXDRbxnTMXDRYOFg4KGgsZCxoKGgsZCxkLGgoZCxoLGQsZCxkXDgoZCxoXDRcNFg4WDhYOChoKGgsZCxkXDQ$
#tendasinistra/down = scCwBOyeBgAFBwQEBEYLBARKBAQESQsEB0YQSQ9IBAgETAQEBEwNBAREBgQEU58yFw0XDRcNCxkLGQsaCxkLGQsZCxkMGQsZCxkLGQwYFw4LGQsZFw0XDRcNFw0WDgsZCxkLGgoaFg4KGgsZFw0LGQsaChoKGhcNCxkLGQsZF/GdMxYOFg4WDgoaChoLGQsZCxoK$
tendasinistra/stop = scASAeKeBgAECgUSBBIGDwYWBAABhgQvBA8EGwQ3njIYDBgMGA0LGQsZCxkLGQsaCxkLGQsZCxkLGQwZCxkXDQsZCxkXDRcNFw0XDRcNCxoKGgoaCxkXDQsZCxkXDQsaChoWDgoaFg4LGRcNCxkX8ZwzFw0XDhYOChoKGgoaChoLGQsZCxoKGgsZCxkLGQsaFg0MG$


[devices]
# Known devices that will be exposed via the bridge.
# The key is the alias that the device can be addressed as.
# The value is the host (name or IP) of the Broadlink device.
myalias = 192.168.1.66

[discovery]
# The bridge can also discover nearby devices (same subnet).
#
# Number of seconds in which to wait for devices to reply during discovery.
# A value of 0 disables auto-discovery.
timeout = 5

[http]
# The port that the REST service will use.
# 8780 is the default port. Use 0 to disable.
port = 8780

[lirc]
# The port that the LIRC service will use. 0 to disable.
# 8765 is the default port. Use 0 to disable.
port = 8765

[mqtt]
# The URL of the MQTT broker in the form:
# <protocol>://[<user>@<password>@]<host>[:<port>]
# where:
# - protocol: 'mqtt' or 'mqtts' (TLS)
# - user (optional)
# - password (optional)
# - host: hostname/IP address of the MQTT server
# - port (optional)
# When empty or missing, the MQTT client is disabled.
broker_url = mqtt://192.168.1.2
# The prefix to use. Default is "broadlink"
```

**Comment by lbschenkel** [comment]
I noticed that the upstream `broadlink` library (that does virtually all the work) had a number of fixes. I just updated this project to use the latest version (0.1.6). Can you check out latest `master` and try again?

**Comment by bubez81** [comment]
sadly same error:
```
2020-11-12 16:45:07,642 INFO     Starting broadlink-bridge/0.1.6...
2020-11-12 16:45:07,642 INFO     Reading config file: config.ini
2020-11-12 16:45:07,729 INFO     Device: DC-EC-F0-A7-DF-24@192.168.1.66 has alias myalias
2020-11-12 16:45:07,729 INFO     Discovery: searching for devices for 5 seconds...
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/bin/broadlink-bridge", line 33, in <module>
    sys.exit(load_entry_point('broadlink-bridge==0.1.6', 'console_scripts', 'broadlink-bridge')())
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/cli.py", line 64, in main
    REGISTRY.discover(config.getint('discovery', 'timeout'))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 30, in discover
    dev = self._add_device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 76, in _add_device
    dev = Device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 99, in __init__
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 107, in connect
    self._dev = broadlink.hello(host=self.host)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 135, in hello
    return next(xdiscover(timeout, local_ip_address, host, port))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 166, in xdiscover
    for resp in responses:
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/device.py", line 78, in scan
    conn.sendto(packet, (discover_ip_address, discover_ip_port))
TypeError: str, bytes or bytearray expected, not rm4
```

**Comment by lbschenkel** [comment]
I'm not sure how I can troubleshoot this further. It's possible the bug is in this project, but I can't reproduce it with the hardware I have (RM3) and I reviewed the source code based on the stack trace and I can't spot anything wrong. What I suggest you to do is to try to play with https://github.com/mjg59/python-broadlink directly and see if you can talk to your hardware. Please report your findings and we can take it from there.

**Comment by bubez81** [comment]
ok!....with the discovery.py no problem:

```
 python3 pprova.py 
Discovering...
###########################################
RM4
# broadlink_cli --type 0x6026 --host 192.168.1.66 --mac 24dfa7f0ecdc
Device file data (to be used with --device @filename in broadlink_cli) : 
0x6026 192.168.1.66 24dfa7f0ecdc
temperature = 0.0
```

**Comment by lbschenkel** [comment]
Hi @bubez81, can you still reproduce this problem with the current version of the bridge?

**Comment by bubez81** [comment]
hello this is the error now:
```2021-09-26 11:28:00,948 INFO     Starting broadlink-bridge/0.1.8...
2021-09-26 11:28:00,948 INFO     Reading config file: config.ini
2021-09-26 11:28:01,036 INFO     Device: 24-DF-A7-F0-EC-DC@192.168.1.66 has alias myalias
2021-09-26 11:28:01,036 INFO     Discovery: searching for devices for 5 seconds...
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/bin/broadlink-bridge", line 33, in <module>
    sys.exit(load_entry_point('broadlink-bridge==0.1.8', 'console_scripts', 'broadlink-bridge')())
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/cli.py", line 89, in main
    REGISTRY.discover(config.getint('discovery', 'timeout'))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 40, in discover
    dev = self._add_device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 86, in _add_device
    dev = Device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 109, in __init__
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 117, in connect
    self._dev = broadlink.hello(host=self.host)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 149, in hello
    return next(xdiscover(timeout, local_ip_address, host, port))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 176, in xdiscover
    for resp in responses:
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/device.py", line 51, in scan
    conn.sendto(packet, (discover_ip_address, discover_ip_port))
TypeError: str, bytes or bytearray expected, not rm4pro
```

**Comment by lbschenkel** [comment]
@bubez81: thanks. do you mind changing `broadlink_bridge/init.py`, line 117, so it changes from this:
```python
                self._dev = broadlink.hello(host=self.host)
```
into this:
```python
                print(self._host)
                print(self._dev)
                print(self._mac)
                self._dev = broadlink.hello(host=self.host)
```
and try it again and send me the output?

There's something different that happens with newer devices but as I don't have any it's hard for me to reproduce the problem, so I need to see what is assigned those variables in your case.

**Comment by bubez81** [comment]
Sure!:
```
2021-09-27 08:19:35,199 INFO     Starting broadlink-bridge/0.1.8...
2021-09-27 08:19:35,199 INFO     Reading config file: /Users/michele/broadlink-bridge/config.ini
192.168.1.66
None
None
2021-09-27 08:19:35,473 INFO     Device: 24-DF-A7-F0-EC-DC@192.168.1.66 has alias myalias
2021-09-27 08:19:35,473 INFO     Discovery: searching for devices for 5 seconds...
Broadlink Soggiorno (Broadlink RM4 pro 0x6026 / 192.168.1.66:80 / 24:DF:A7:F0:EC:DC)
None
None
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/bin/broadlink-bridge", line 33, in <module>
    sys.exit(load_entry_point('broadlink-bridge==0.1.8', 'console_scripts', 'broadlink-bridge')())
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/cli.py", line 89, in main
    REGISTRY.discover(config.getint('discovery', 'timeout'))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 40, in discover
    dev = self._add_device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 86, in _add_device
    dev = Device(dev)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 109, in __init__
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink_bridge/__init__.py", line 120, in connect
    self._dev = broadlink.hello(host=self.host)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 149, in hello
    return next(xdiscover(timeout, local_ip_address, host, port))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/__init__.py", line 176, in xdiscover
    for resp in responses:
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/broadlink/device.py", line 51, in scan
    conn.sendto(packet, (discover_ip_address, discover_ip_port))
TypeError: str, bytes or bytearray expected, not rm4pro
```

**Comment by chkuendig** [comment]
I'm encountering this issue with the rm4mini on the CLI as well. It's not clear to me whether this is a bug in broadlink-bridge or the hello/xdiscovery method of python-broadlink (python-broadlink via its CLI works for me, including learning, sending - but i don't think that CLI uses hello/xdiscovery)

**Comment by kathampy** [comment]
I have the same error with the RM4 Mini.
`TypeError: str, bytes or bytearray expected, not rm4mini`

**Comment by lbschenkel** [comment]
Is there a developer in the audience? Since I don't have the hardware and it works on mine it's very difficult for me to troubleshoot and fix the problem "blind" (unless someone wants to donate some hardware for me to test against).

For the technically inclined, what I suggest is to upgrade `python-broadlink` and if that still doesn't work, play with the code and figure out what's breaking. With a detailed explanation of the problem I can then fix the code.

**Comment by kathampy** [comment]
I've never used Python before. How do I rebuild this package with locally modified files in the `venv`'s `Lib\site-packages\broadlink_bridge` folder? I originally installed it using the `pip` command.

**Comment by lbschenkel** [comment]
> I've never used Python before. How do I rebuild this package with locally modified files in venv `Lib\site-packages\broadlink_bridge`? I originally installed it using the `pip` command.

In that case you're not the audience I have in mind (no offense). What I'm looking for is somebody who is affected and knows how to play with the codebase and do some experiments (change code, do some debugging).

**Comment by chkuendig** [comment]
I have the skills to look into this. I'll see if I find some time in the coming days. Updating python-broadlink unfortunatey didnt help.

**Comment by lbschenkel** [comment]
> I have the skills to look into this. I'll see if I find some time in the coming days. Updating broadink-python unfortunatey didnt help.

Ideally I should use the same method as `python-broadlink` CLI and avoid any other less tested code paths, but last time I looked this was not powerful enough for the purposes of the bridge (for example: connecting to a well-known host so it can be in a different broadcast domain) and I was forced to use the internal API. Maybe I should look again and this has changed.

**Comment by kathampy** [comment]
How do I run this from the source directory without `setup.py install`ing the whole module? What is the command to directly launch the main method? Is that even possible?

**Comment by kathampy** [comment]
`python -m broadlink_bridge.cli` return blank without any error or output.

**Comment by kathampy** [comment]
Finally got it to run locally. What a terrible development environment this language has...
```
git clone https://github.com/lbschenkel/broadlink-bridge.git
python -m venv .venv
.venv\Scripts\activate
pip install -e broadlink-bridge
broadlink-bridge config.ini
```

**Comment by kathampy** [comment]
Submitted a PR. The `dev` argument was being incorrectly passed to the `host` parameter in the `Device` constructor. It runs now, but with a logging error due to `mac` being `None`.
```
2022-01-14 14:39:55,687 INFO     Starting broadlink-bridge/0.1.8...
2022-01-14 14:39:55,688 INFO     Discovery: searching for devices for 5 seconds...
--- Logging error ---
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.496.0_x64__qbz5n2kfra8p0\lib\logging\__init__.py", line 1098, in emit
    msg = self.format(record)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.496.0_x64__qbz5n2kfra8p0\lib\logging\__init__.py", line 942, in format
    return fmt.format(record)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.496.0_x64__qbz5n2kfra8p0\lib\logging\__init__.py", line 678, in format
    record.message = record.getMessage()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.496.0_x64__qbz5n2kfra8p0\lib\logging\__init__.py", line 368, in getMessage
    msg = msg % self.args
  File "c:\users\kurian\broadlink-bridge\broadlink_bridge\__init__.py", line 182, in __str__
    return self.mac + '@' + self.host
TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'
Call stack:
  File "C:\Users\Kurian\blvenv\Scripts\broadlink-bridge-script.py", line 33, in <module>
    sys.exit(load_entry_point('broadlink-bridge', 'console_scripts', 'broadlink-bridge')())
  File "c:\users\kurian\broadlink-bridge\broadlink_bridge\cli.py", line 89, in main
    REGISTRY.discover(config.getint('discovery', 'timeout'))
  File "c:\users\kurian\broadlink-bridge\broadlink_bridge\__init__.py", line 40, in discover
    dev = self._add_device(dev)
  File "c:\users\kurian\broadlink-bridge\broadlink_bridge\__init__.py", line 91, in _add_device
    LOGGER.info('Device: %s has alias %s', dev, alias)
Unable to print the message and arguments - possible formatting error.
Use the traceback above to help find the error.
2022-01-14 14:40:00,726 INFO     HTTP server started on port 8780
2022-01-14 14:40:00,728 INFO     LIRC server started on port 8765
2022-01-14 14:40:00,728 INFO     MQTT client disabled
```

**Comment by kathampy** [comment]
Updated the PR. There was another issue where `auth()` would not be called when `Device` was constructed with `dev` instead of a `host` string. Fixing that also resolved the above logging error.

It seems to be working fine with my `RM4 mini` now. I can send codes via HTTP.

**Comment by lbschenkel** [comment]
This should be fixed by #15. Please try the `master` branch again and it's not working for you, please reopen.
